fix(mme): PDUSessionResourceSetupRequest failure randomly

ganeshg87 opened the pull request:
Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary
  fix(mme): PDUSessionResourceSetupRequest failure randomly

## Test Plan

 below flows verified using UERANSIM
      1. registration setup
      2. pdu session
      3. deregistration


## Additional Information

1. pdu session resource setup request failed due to number active pdusession variable is uninitialized 
![bb87a45ae1c55486f51997eaa008cf6f46ae3f6f220e88b4b145572bc333e30f](https://user-images.githubusercontent.com/83060027/128134123-99dcb98c-48c6-4081-aaa6-6283e2fa0c14.png)
2. pdu session resource setup request failed due to pds_id.guti  variable is uninitilized 
![image](https://user-images.githubusercontent.com/83060027/128134453-469aac9a-5a08-4298-bce3-044d0567bab1.png)

ganeshg87 committed:
fix(mme): PDUSessionResourceSetupRequest failure randomly

issue:
    pdu session setup failing due to initilized context variables

test cases:
    verified below cases with UERANSIM
    1. registration
    2. pdusession
    3. deregistration

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

ulaskozat commented on lte/gateway/c/core/oai/tasks/amf/amf_as.cpp:
```diff
@@ -1191,11 +1191,6 @@ uint16_t amf_as_establish_cnf(
-    OAILOG_WARNING(LOG_NAS_AMF, "AMFAS-SAP - GUTI is NULL...");
-    OAILOG_FUNC_RETURN(LOG_NAS_AMF, ret_val);
-  }
-
```
why was this check here before and why is it removed now?

ganeshg87 replied:
The above code was committed from Chandra on April 12(#5984).
reviewed PR, above specific variable, is not initialized.

when I was using the UERANSIM script with Registration & PDUSession, sometimes it was failing due to the above check.
backtraced pds_id.guti variable is unused & uninitilzed value.

ganeshg87 replied:
till now we did not see any issue because pds_id.guti value garbage, and amf not used this variable.

ganeshg87 replied:
basically, amf_sap structure is not initialized,  I have called memset all places where  amf_sap  defined

ganeshg87 replied:
function local variables initialization is a basic coding guideline, it was missed in base code.
I did not change any functionality and not added any UT cases.
 
Please provide your feedback,

ulaskozat commented on lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp:
```diff
@@ -139,6 +139,7 @@ ue_m5gmm_context_s* amf_create_new_ue_context(void) {
     OAILOG_ERROR(LOG_AMF_APP, "Failed to allocate memory for UE context \n");
     return NULL;
   }
+  memset(new_p, 0, sizeof(ue_m5gmm_context_s));
```
add unit test coverage for `amf_create_new_ue_context`

ganeshg87 replied:
there was no issue in  ue_m5gmm_context_s initialization, safer side I added.
but ue_m5gmm_context_s  struct   allocated with c++ constructor(new ue_m5gmm_context_s()  ),properly all members are initialized, removed memset and validated with UT.

ulaskozat commented on lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp:
```diff
@@ -217,6 +218,7 @@ ue_m5gmm_context_s* amf_ue_context_exists_amf_ue_ngap_id(
 smf_context_t* amf_insert_smf_context(
     ue_m5gmm_context_s* ue_context, uint8_t pdu_session_id) {
   smf_context_t smf_context;
+  memset(&smf_context, 0, sizeof(smf_context));
```
add unit test coverage for `amf_insert_amf_context`

ganeshg87 replied:
comment addressed added test coverage.

ganeshg87 committed:
fix(mme): addressed review comments, additional changes related initilization

test cases:
 - basic sanity done with UERANSIM
   basic registartion
   pdu session setup
   deregistration
 - verified UT modules

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

ganeshg87 commented:
- basic sanity done with UERANSIM
           initial registration, pdusesssion, deregister flows.
![image](https://user-images.githubusercontent.com/83060027/128491502-7c814fda-92c7-4094-a1fd-f079a0edcb21.png)
[mme_ue_register_06Aug2021_092923.log](https://github.com/magma/magma/files/6944660/mme_ue_register_06Aug2021_092923.log)

- added new UT cases and executed UT modules

![image](https://user-images.githubusercontent.com/83060027/128491714-0b5a3f52-89ff-4e31-b99e-7e98d4bc6d70.png)

ganeshg87 commented:
I have replaced memset with C++11 '{}' as per Yogesh review comment.

ulaskozat commented on lte/gateway/c/core/oai/test/amf/test_amf.cpp:
```diff
@@ -10,8 +10,28 @@
+  {THREAD_##tHREADiD, #tHREADiD, "ipc://IPC_" #tHREADiD},
+#include <tasks_def.h>
+#undef TASK_DEF
+};
```
I do not see any itti messages or tasks being used in these tests. What is the reason for including these?

ganeshg87 replied:
addressed review comment removed unused itti header files.

ganeshg87 committed:
fix(mme):addressed review comments

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

ganeshg87 commented:
- verified oai UT cases
![image](https://user-images.githubusercontent.com/83060027/128727462-a130e89a-a4df-4444-8c4f-2a4518257cc0.png)

ulaskozat commented on lte/gateway/c/core/oai/test/amf/test_amf.cpp:
```diff
@@ -10,11 +10,15 @@
 
 using ::testing::Test;
 
+task_zmq_ctx_t grpc_service_task_zmq_ctx;
```
is this used? If not remove it.

ganeshg87 replied:
n11 has a dependency with grpc_service, it requires  grpc_service_task_zmq_ctx, If I remove the above, getting the below error due to dependency.
![image](https://user-images.githubusercontent.com/83060027/128896577-9e46328e-f9b7-40a1-8f86-49bdd2f5a77f.png)

ulaskozat commented on lte/gateway/c/core/oai/test/amf/test_amf.cpp:
```diff
@@ -10,11 +10,15 @@
 #include "util_nas5g_pkt.h"
-#include "dynamic_memory_check.h"
 #include <gtest/gtest.h>
+#include "intertask_interface.h"
```
is this header used? If not remove it.

ganeshg87 replied:
grpc_service_task_zmq_ctx was declared in intertask_interface.h

ardzoht reviewed (approved):
lgtm

ulaskozat reviewed (approved):
LGTM
